File: backend/FlashcardsApi/Services/QdrantLearningPathService.cs
using FlashcardsApi.Models;
using Qdrant.Client;
using Qdrant.Client.Grpc;
using System.Text.Json;

namespace FlashcardsApi.Services
{
    public class QdrantLearningPathService : ILearningPathService
    {
        private readonly QdrantClient _client;
        private readonly string _collectionName = "learning_paths";
        private readonly int _vectorSize = 64; // Adjust as needed

        public QdrantLearningPathService(string host, int port)
        {
            _client = new QdrantClient(host, port);
        }

        public async Task<IEnumerable<LearningPath>> GetAllAsync()
        {
            var result = new List<LearningPath>();
            PointId? offset = null;
            const int limit = 1000;
            while (true)
            {
                var scrollRes = await _client.ScrollAsync(_collectionName, limit: limit, offset: offset);
                foreach (var point in scrollRes.Result)
                {
                    if (point.Payload != null && point.Payload.TryGetValue("json", out var jsonVal))
                    {
                        var path = JsonSerializer.Deserialize<LearningPath>(jsonVal.ToString());
                        if (path != null) result.Add(path);
                    }
                }
                if (scrollRes.Result.Count < limit) break;
                offset = scrollRes.NextPageOffset;
            }
            return result;
        }

        public async Task AddAsync(LearningPath path)
        {
            var vector = new float[_vectorSize];
            var point = new PointStruct
            {
                Id = new PointId { Uuid = path.Id },
                Vectors = vector,
                Payload = { ["json"] = JsonSerializer.Serialize(path) }
            };
            await _client.UpsertAsync(_collectionName, new List<PointStruct> { point });
        }

        public async Task UpdateAsync(LearningPath path)
        {
            await AddAsync(path);
        }

        public async Task DeleteAsync(string id)
        {
            var filter = new Filter
            {
                Must = { new Condition
                    {
                        HasId = new HasIdCondition
                        {
                            HasId = { new PointId { Uuid = id } }
                        }
                    }
                }
            };
            await _client.DeleteAsync(_collectionName, filter);
        }

        public async Task<LearningPath?> GetByIdAsync(string id)
        {
            var filter = new Filter
            {
                Must = { new Condition
                    {
                        HasId = new HasIdCondition
                        {
                            HasId = { new PointId { Uuid = id } }
                        }
                    }
                }
            };
            var result = await _client.ScrollAsync(_collectionName, filter: filter, limit: 1);
            var point = result.Result.FirstOrDefault();
            if (point?.Payload != null && point.Payload.TryGetValue("json", out var jsonVal))
            {
                return JsonSerializer.Deserialize<LearningPath>(jsonVal.ToString());
            }
            return null;
        }

        public async Task<(bool Success, string Message)> SeedFromJsonAsync()
        {
            var path = Path.Combine(AppContext.BaseDirectory, "Resources", "learning-paths.json");
            if (!System.IO.File.Exists(path))
                return (false, "learning-paths.json not found");
            var json = await System.IO.File.ReadAllTextAsync(path);
            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            var paths = JsonSerializer.Deserialize<List<LearningPath>>(json, options);
            if (paths == null) return (false, "Invalid JSON content");
            foreach (var pathItem in paths)
            {
                if (string.IsNullOrWhiteSpace(pathItem.Id))
                    pathItem.Id = Guid.NewGuid().ToString();
                await AddAsync(pathItem);
            }
            return (true, $"{paths.Count} learning paths seeded.");
        }
    }
}

File: backend/app/services/qdrant_learning_path_service.py
from __future__ import annotations
from typing import List, Iterable, Tuple, Optional
from qdrant_client import QdrantClient
from qdrant_client.http.models import (
    PointStruct,
    VectorParams,
    Distance,
    Filter,
    HasIdCondition,
)

try:
    from qdrant_client.http.models import PointId
except ImportError:  # Older or newer qdrant-client versions may not expose PointId
    PointId = str  # type: ignore
from ..models import LearningPath
import uuid
import os
import random
import json


class QdrantLearningPathService:
    def __init__(
        self,
        host: str = "10.0.0.9",
        port: int = 6334,
        collection: str = "learning_paths",
        vector_size: int = 64,
    ):
        # The provided port corresponds to the gRPC interface.  Configure the
        # client to use it instead of the REST API.
        self.client = QdrantClient(host=host, grpc_port=port, prefer_grpc=True)
        self.collection = collection
        self.vector_size = vector_size
        self._ensure_collection()

    def _ensure_collection(self):
        # Ensure the collection exists.  If it does not, create it with the
        # specified vector size and distance metric.
        # Note: This does not check for existing collections with the same name
        # but rather creates a new one if it does not exist.
        existing = self.client.get_collections().collections
        names = [c.name for c in existing]
        if self.collection not in names:
            self.client.create_collection(
                collection_name=self.collection,
                vectors_config=VectorParams(
                    size=self.vector_size, distance=Distance.COSINE
                ),
            )

    def add(self, path: LearningPath):
        """Add a learning path to the Qdrant collection."""
        # If the path does not have an ID, generate a new one.
        # If the ID is in the form of a dictionary, extract the UUID.
        # Otherwise, use the existing ID.

        if not path.id:
            path.id = str(uuid.uuid4())
        elif isinstance(path.id, dict) and "uuid" in path.id:
            path.id = str(path.id["uuid"])
        vector = [0.0] * self.vector_size
        # Ensure the point id is a simple string before sending it to Qdrant.
        point = PointStruct(
            id=str(path.id), vector=vector, payload={"json": path.json()}
        )
        self.client.upsert(collection_name=self.collection, points=[point])

    def update(self, path: LearningPath):
        self.add(path)

    def delete(self, path_id: str):
        flt = Filter(must=[HasIdCondition(has_id=[PointId(str(path_id))])])
        self.client.delete(collection_name=self.collection, points_selector=flt)

    def get_all(self) -> List[LearningPath]:
        points = self.client.scroll(collection_name=self.collection, limit=1000)[0]
        paths: List[LearningPath] = []
        for p in points:
            if p.payload and "json" in p.payload:
                data = json.loads(p.payload["json"])
                paths.append(LearningPath(**data))
        return paths

    def get_by_id(self, path_id: str) -> Optional[LearningPath]:
        """Retrieve a learning path by its ID."""
        points = self.client.scroll(
            collection_name=self.collection,
            limit=1,
            filter=Filter(must=[HasIdCondition(has_id=[PointId(str(path_id))])]),
        )[0]
        for p in points:
            if p.payload and "json" in p.payload:
                data = json.loads(p.payload["json"])
                return LearningPath(**data)
        return None

    def seed_from_json(self, path: str) -> Tuple[bool, str]:
        """Seed the Qdrant collection with learning paths from a JSON file."""

        if not os.path.exists(path):
            return False, "learning-paths.json not found"
        with open(path, "r", encoding="utf-8") as f:
            paths = json.load(f)
        if not isinstance(paths, Iterable):
            return False, "Invalid JSON content"
        count = 0
        for data in paths:
            lp = LearningPath(**data)
            if not lp.id:
                lp.id = str(uuid.uuid4())
            self.add(lp)
            count += 1
        return True, f"{count} learning paths seeded."
